### ISSUE TITLE
n() no longer works if dplyr not attached

### ISSUE THREAD

**Issue by JohnMount** [tidyverse/dplyr#4062]
``` r
packageVersion("dplyr")
#> [1] '0.7.99.9000'

# fails 
dplyr::mutate(data.frame(x=1), count = n())
#> Error in n(): could not find function "n"

# works
dplyr::mutate(data.frame(x=1), count = dplyr::n())
#>   x count
#> 1 1     1

# works
library("dplyr")
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
dplyr::mutate(data.frame(x=1), count = n())
#>   x count
#> 1 1     1
```

<sup>Created on 2018-12-31 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

I am filing this as requested in the email I received today, excerpted below.

> This is an automated email to let you know that:
> * A new version of dplyr is ready to go to CRAN. dplyr is currently at version 0.7.8 and will become 0.8.0 upon release.
> If you have discovered a bug in dplyr, please file an issue (ideally with a small reprex that illustrates the problem) at

By my count this change breaks about 70 packages.  Also it seems the proposed `dplyr` update is breaking about 120 of 705 dependent packages (about 17%).

**Comment by joranE** [comment]
I believe this was intentional, e.g. [#4030](https://github.com/tidyverse/dplyr/issues/4030)

**Comment by JohnMount** [comment]
There is no way to work around the issue for a database:

``` r
packageVersion("dplyr")
#> [1] '0.7.8'

# works
dplyr::summarize(data.frame(x = 1), count = dplyr::n())
#>   count
#> 1     1

# fails
db <- DBI::dbConnect(RSQLite::SQLite(), ":memory:")
t <- dplyr::copy_to(db, data.frame(x = 1), "d")
dplyr::summarize(t, count = dplyr::n())
#> Error: This function should not be called directly
```

<sup>Created on 2018-12-31 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by JohnMount** [comment]
Also the `replyr` package had the same issue, but this was not caught in the check.  So there may be many more packages affected.

**Comment by hadley** [comment]
Yes, this is a breaking change that will affect code. We are doing it to make dplyr consistent with usual R semantics. As with all breaking changes, it causes some short term pain in return for pain reduction in the long term.

**Comment by yutannihilation** [comment]
I agree this is worth breaking. It is error-prone to allow unqualified `n()` without attaching dplyr. For example:

``` r
n <- function() stop("foo")
dplyr::summarise(iris, n())
#> Warning: The `printer` argument is soft-deprecated as of rlang 0.3.0.
#> This warning is displayed once per session.
#> Error in summarise_impl(.data, dots): Evaluation error: foo.
```

<sup>Created on 2019-01-01 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

Actually, I found #4030 when I was investigating the problem caused by the conflict between dplyr's `n()` and SparkR's `n()`, which took time to figure out what happened... So, I welcome this breaking change 👍

**Comment by lionel-** [comment]
@yutannihilation If you install latest dev rlang, you shouldn't see these warnings about `printer` anymore. I have fixed a bug in the code that decides when to warn.

**Comment by yutannihilation** [comment]
@lionel- Thanks, I'll try the dev version! (FYI: I haven't seen this warning with the dev version of dplyr. The reprex above was created with dplyr 0.7.8.)
